assets/src/edit-story/app/history/test/reducer.js

```diff
@@ -0,0 +1,273 @@
+        payload: {},
+      })
+    ).toThrow(/Unknown history reducer action: UNKNOWN_ACTION/i);
+  });
+
+  describe('addEntry', () => {
+    it('should add entry to the beginning of the history', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 1 }],
+      };
+
+      const newEntry = { id: 2 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.entries[0]).toStrictEqual(newEntry);
+    });
+
+    it('should clear undone states when adding a new entry', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 1 }, { id: 2 }, { id: 3 }],
+        offset: 2,
+      };
+      const newEntry = { id: 4 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.entries).toHaveLength(2);
+      expect(result.entries[1].id).toStrictEqual(3);
+      expect(result.entries[0].id).toStrictEqual(4);
+    });
+
+    it('should set correct state values when adding new entry', () => {
+      const initialState = {
+        ...EMPTY_STATE,
+        versionNumber: 3,
+        entries: [{ id: 1 }],
+        offset: 2,
+      };
+      const newEntry = { id: 2 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+
+      expect(result.versionNumber).toStrictEqual(4);
+      expect(result.offset).toStrictEqual(0);
+    });
+
+    it('should respect the determined history size when adding a new entry', () => {
+      const entries = [...Array(size + 1)].map(() => {
+        return { id: 'foo' };
+      });
+      const initialState = {
+        ...EMPTY_STATE,
+        entries,
+      };
+      const newEntry = { id: 7 };
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: newEntry,
+      });
+      expect(result.entries).toHaveLength(size);
+    });
+
+    it('should set the correct offset when replaying', () => {
+      const requestedState = { id: 1 };
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 2 }, requestedState],
+        requestedState,
+      };
+
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: requestedState,
+      });
+
+      expect(result.offset).toStrictEqual(1);
+    });
+
+    it('should clear the requested state after replaying', () => {
+      const requestedState = { id: 1 };
+      const initialState = {
+        ...EMPTY_STATE,
+        entries: [{ id: 2 }, requestedState],
+        requestedState,
+      };
+
+      const result = reducer(initialState, {
+        type: SET_CURRENT_STATE,
+        payload: requestedState,
+      });
+
+      expect(result.requestedState).toBeNull();
+    });
+
+    it('should set replay the current page relevant to the latest replayed change', () => {
```

Comment:
What's the purpose of this test again?

Reply:
To test dynamically setting the current page so that the user would always stay on the page where the relevant change happened.

E.g. if the user is on page 1, then goes to page 2, makes a change, and clicks undo, even though the previous replay state has current: 1, we'll display 2 dynamically for staying on the same page.

This test basically tests that this would actually happen.

Reply:
It's just that the test description is hard to read. Maybe some missing words or maybe some half-completed editing?

Reply:
Oops, the description is very confusing indeed. Changed now in fb0b4539.